File: cpp/lib/split.cpp
#include "split.hpp"

namespace Blastn {

vector<string> split_to_words(string& input, u32 word_len) {
    u32 num_words = (u32)input.length() - word_len + 1;
    vector<string> words;
    words.reserve(num_words);

    // get all substrings of the same length in the input
    for (u32 i = 0; i < num_words; i++)
        words.emplace_back(input.substr(i, word_len));

    return words;
}

} // Blastn

File: python/lib/split.py
import sys

"""
External
"""

def split_to_words(iterable: iter, length: int) -> list:
    """
    @brief: Split an iterable into all possible slices of given length \\
    @param iterable: A Python iterable to split into a list of words of length \\
    @param length:   The length of the words to extract from the iterable \\
    @return: A list of words of given length split from the iterable
    """
    num_words: int = len(iterable) - length + 1
    words: list = []
    for i in range(num_words):
        words.append(iterable[i:i + length])
    
    return words
